hazelcast/protocol/builtin.py

```diff
@@ -496,3 +554,203 @@ def encode(buf, value, is_final=False):
+
+
+class ListCNLocalDateTimeCodec(object):
+    @staticmethod
+    def decode(msg):
+        return ListCNFixedSizeCodec.decode(
+            msg, _LOCAL_DATE_TIME_SIZE_IN_BYTES, FixSizedTypesCodec.decode_local_date_time
+        )
+
+
+class ListCNOffsetDateTimeCodec(object):
+    @staticmethod
+    def decode(msg):
+        return ListCNFixedSizeCodec.decode(
+            msg, _OFFSET_DATE_TIME_SIZE_IN_BYTES, FixSizedTypesCodec.decode_offset_date_time
+        )
+
+
+class BigDecimalCodec(object):
+    @staticmethod
+    def decode(msg):
+        buf = msg.next_frame().buf
+        size = FixSizedTypesCodec.decode_int(buf, 0)
+        unscaled_value = int_from_bytes(buf[INT_SIZE_IN_BYTES : INT_SIZE_IN_BYTES + size])
+        scale = FixSizedTypesCodec.decode_int(buf, INT_SIZE_IN_BYTES + size)
+        sign = 0 if unscaled_value >= 0 else 1
+        return Decimal((sign, tuple(int(digit) for digit in str(abs(unscaled_value))), -1 * scale))
+
+
+class SqlPageCodec(object):
+    @staticmethod
+    def decode(msg):
+        from hazelcast.sql import SqlColumnType, _SqlPage
+
+        # begin frame
+        msg.next_frame()
+
+        # read the "last" flag
+        is_last = LE_INT8.unpack_from(msg.next_frame().buf, 0)[0] == 1
+
+        # read column types
+        column_type_ids = ListIntegerCodec.decode(msg)
+
+        # read columns
+        columns = []
+
+        for column_type_id in column_type_ids:
+            if column_type_id == SqlColumnType.VARCHAR:
+                columns.append(
+                    ListMultiFrameCodec.decode_contains_nullable(msg, StringCodec.decode)
+                )
+            elif column_type_id == SqlColumnType.BOOLEAN:
+                columns.append(ListCNBooleanCodec.decode(msg))
+            elif column_type_id == SqlColumnType.TINYINT:
+                columns.append(ListCNByteCodec.decode(msg))
+            elif column_type_id == SqlColumnType.SMALLINT:
+                columns.append(ListCNShortCodec.decode(msg))
+            elif column_type_id == SqlColumnType.INTEGER:
+                columns.append(ListCNIntegerCodec.decode(msg))
+            elif column_type_id == SqlColumnType.BIGINT:
+                columns.append(ListCNLongCodec.decode(msg))
+            elif column_type_id == SqlColumnType.REAL:
+                columns.append(ListCNFloatCodec.decode(msg))
+            elif column_type_id == SqlColumnType.DOUBLE:
+                columns.append(ListCNDoubleCodec.decode(msg))
+            elif column_type_id == SqlColumnType.DATE:
+                columns.append(ListCNLocalDateCodec.decode(msg))
+            elif column_type_id == SqlColumnType.TIME:
+                columns.append(ListCNLocalTimeCodec.decode(msg))
+            elif column_type_id == SqlColumnType.TIMESTAMP:
+                columns.append(ListCNLocalDateTimeCodec.decode(msg))
+            elif column_type_id == SqlColumnType.TIMESTAMP_WITH_TIME_ZONE:
+                columns.append(ListCNOffsetDateTimeCodec.decode(msg))
```

Comment:
Where do you convert these to string? In the documentation, you say strings are returned.